### ISSUE TITLE
nn.utils.spectral_norm() does not perform normalization of the weight matrix.

### ISSUE THREAD

**Issue by k-melnyk** [pytorch/pytorch#48181]
## To Reproduce

```python
import torch

linear = torch.nn.Linear(3, 4)
norm_layer = torch.nn.utils.spectral_norm(linear)

print('Normalized weight matrix with spectral_norm(): ', norm_layer.weight)
print('Original weight matrix: ', norm_later.weight_orig)

sigma = torch.dot(norm_layer.weight_u, torch.mv(norm_layer.weight_orig, norm_layer.weight_v))
print('Normalized weight matrix by hands: ', a.weight_orig / sigma)
```
* Spectral_norm() does not normalize the original weight matrix;
* The first two prints will give the same weight matrix;
* If we compute sigma using members weight_u and weight_v and then normalize original weight matrix, it will produce a different weight matrix;

## Environment

 - PyTorch Version (e.g., 1.0): 1.7
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source):
 - Python version: 3.8
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: Nvidia RTX 2060
 - Any other relevant information:

cc @albanD @mruberry

**Comment by SsnL** [comment]
It's implemented as a forward hook. You get normalized w in forward.

**Comment by mruberry** [comment]
I corrected the script:

```
import torch
print(torch.__version__)

linear = torch.nn.Linear(3, 4)
norm_layer = torch.nn.utils.spectral_norm(linear)

print('Normalized weight matrix with spectral_norm(): ', norm_layer.weight)
print('Original weight matrix: ', norm_layer.weight_orig)

sigma = torch.dot(norm_layer.weight_u, torch.mv(norm_layer.weight_orig, norm_layer.weight_v))
print('Normalized weight matrix by hands: ', norm_layer.weight_orig / sigma)
```

It originally prints:

```
1.7.0+cu101
Normalized weight matrix with spectral_norm():  tensor([[ 0.1368,  0.3952,  0.4573],
        [-0.3129,  0.5319,  0.0304],
        [ 0.2035,  0.2506,  0.3884],
        [ 0.4142,  0.0710,  0.4722]])
Original weight matrix:  Parameter containing:
tensor([[ 0.1368,  0.3952,  0.4573],
        [-0.3129,  0.5319,  0.0304],
        [ 0.2035,  0.2506,  0.3884],
        [ 0.4142,  0.0710,  0.4722]], requires_grad=True)
Normalized weight matrix by hands:  tensor([[-0.2751, -0.7945, -0.9191],
        [ 0.6290, -1.0692, -0.0610],
        [-0.4091, -0.5038, -0.7808],
        [-0.8326, -0.1427, -0.9493]], grad_fn=<DivBackward0>)
```

And yes, the first two printed matrices are the same.

As @SsnL suggests, we then have to forward through the module:

```
norm_layer(torch.randn((3)))
```

And printing the tensors afterwards:

```
print('Normalized weight matrix with spectral_norm(): ', norm_layer.weight)
print('Original weight matrix: ', norm_layer.weight_orig)
```

Prints:

```
Normalized weight matrix with spectral_norm():  tensor([[ 0.1502,  0.4337,  0.5017],
        [-0.3433,  0.5837,  0.0333],
        [ 0.2233,  0.2750,  0.4262],
        [ 0.4545,  0.0779,  0.5182]], grad_fn=<DivBackward0>)
Original weight matrix:  Parameter containing:
tensor([[ 0.1368,  0.3952,  0.4573],
        [-0.3129,  0.5319,  0.0304],
        [ 0.2035,  0.2506,  0.3884],
        [ 0.4142,  0.0710,  0.4722]], requires_grad=True)
```

And the matrices are different. 

The normalized weight matrix is not equal to the "by hand" normalization the script is performing, though. There may be a discrepancy between the two algorithms.

**Comment by SsnL** [comment]
It does power iteration at forward, which updates the parameters. If you compute again after forward, the matrices match:
```py
In [1]: import torch
   ...: print(torch.__version__)
   ...:
   ...: linear = torch.nn.Linear(3, 4)
   ...: norm_layer = torch.nn.utils.spectral_norm(linear)
   ...:
   ...: print('Normalized weight matrix with spectral_norm(): ', norm_layer.weight)
   ...: print('Original weight matrix: ', norm_layer.weight_orig)
   ...:
   ...: sigma = torch.dot(norm_layer.weight_u, torch.mv(norm_layer.weight_orig, norm_layer.weight_v))
   ...: print('Normalized weight matrix by hands: ', norm_layer.weight_orig / sigma)
1.7.0a0+33e2665
Normalized weight matrix with spectral_norm():  tensor([[ 0.4658,  0.3948, -0.1485],
        [-0.3340,  0.4385, -0.2675],
        [-0.2579,  0.4863, -0.2420],
        [ 0.3940,  0.4005, -0.3857]])
Original weight matrix:  Parameter containing:
tensor([[ 0.4658,  0.3948, -0.1485],
        [-0.3340,  0.4385, -0.2675],
        [-0.2579,  0.4863, -0.2420],
        [ 0.3940,  0.4005, -0.3857]], requires_grad=True)
Normalized weight matrix by hands:  tensor([[-2.9618, -2.5104,  0.9443],
        [ 2.1241, -2.7887,  1.7012],
        [ 1.6401, -3.0922,  1.5386],
        [-2.5056, -2.5471,  2.4527]], grad_fn=<DivBackward0>)

In [2]: norm_layer(torch.randn((3)))
Out[2]: tensor([-0.5175,  0.6595,  1.0445,  0.0289], grad_fn=<AddBackward0>)

In [3]: print('Normalized weight matrix with spectral_norm(): ', norm_layer.weight)
   ...: print('Original weight matrix: ', norm_layer.weight_orig)
Normalized weight matrix with spectral_norm():  tensor([[ 0.4720,  0.4001, -0.1505],
        [-0.3385,  0.4444, -0.2711],
        [-0.2614,  0.4928, -0.2452],
        [ 0.3993,  0.4059, -0.3909]], grad_fn=<DivBackward0>)
Original weight matrix:  Parameter containing:
tensor([[ 0.4658,  0.3948, -0.1485],
        [-0.3340,  0.4385, -0.2675],
        [-0.2579,  0.4863, -0.2420],
        [ 0.3940,  0.4005, -0.3857]], requires_grad=True)

In [4]: sigma = torch.dot(norm_layer.weight_u, torch.mv(norm_layer.weight_orig, norm_layer.weight_v))
   ...: print('Normalized weight matrix by hands: ', norm_layer.weight_orig / sigma)
Normalized weight matrix by hands:  tensor([[ 0.4720,  0.4001, -0.1505],
        [-0.3385,  0.4444, -0.2711],
        [-0.2614,  0.4928, -0.2452],
        [ 0.3993,  0.4059, -0.3909]], grad_fn=<DivBackward0>)

```

**Comment by mruberry** [comment]
I believe that closes this issue. The documentation also describes this behavior.
